apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDataReader.cs

```diff
@@ -215,16 +215,52 @@ public Value GetJsonValue(int i)
         /// <inheritdoc />
         public override int GetOrdinal(string name)
         {
-            // If we've already populated the field index, complete synchronously.
+            // We can complete synchronously so long as we have the metadata already.
+            // (If we haven't already populated the field index, we can do so synchronously if there's metadata.)
+            // The only time we *wouldn't* already have the metadata is if Read/ReadAsync hasn't already been
+            // called. That isn't as uncommon as it sounds though, as it's not unreasonable to get the ordinals
+            // of columns before entering a loop.
             GaxPreconditions.CheckNotNullOrEmpty(name, nameof(name));
-            if (_fieldIndex.Count != 0)
+            if (_fieldIndex is null)
             {
-                return _fieldIndex[name];
+                PopulateMetadata();
+                InitializeFieldIndexFromMetadata();
             }
-            // Otherwise, fetch metadata in a new task and wait for it.
-            return Task.Run(() => GetFieldIndexAsync(name, CancellationToken.None)).ResultWithUnwrappedExceptions();
+            return _fieldIndex[name];
         }
-        
+
+        /// <summary>
+        /// Returns the column ordinal given the name of the column, asynchronously requesting metadata from
+        /// the database if necessary.
+        /// </summary>
+        /// <param name="name">The name of the column.</param>
+        /// <param name="cancellationToken">A cancellation token to cancel reading metadata. Defaults to <see cref="CancellationToken.None"/>.</param>
+        /// <returns>The zero-based column ordinal.</returns>
+        public async Task<int> GetOrdinalAsync(string name, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNullOrEmpty(name, nameof(name));
+            if (_fieldIndex is null)
+            {
+                await PopulateMetadataAsync(cancellationToken).ConfigureAwait(false);
+                InitializeFieldIndexFromMetadata();
+            }
+            return _fieldIndex[name];
+        }
+
+        private void InitializeFieldIndexFromMetadata()
+        {
+            _fieldIndex = new ConcurrentDictionary<string, int>();
```

Comment:
Consider moving this line inside the if. As it stands, if this method is called without having called PopulateMetada, the index collection gets initialized but not populated, and that means that it won't ever get populated.
Or if this method should always be called after having metadata populated, just throw if thre's no metadata.

Reply:
I'm not sure it should throw though. In at least some places, it looks like it's valid for there to be no metadata. We're not consistent about handling that, but I'd like to avoid failing if that's the case (when we can let the GetOrdinal method through due to an invalid key instead).

How about I just add a comment above the method saying that PopulateMetadata (or async) must already have been called?

Reply:
Yep, OK, that's fine.